numba/np/arrayobj.py

```diff
@@ -2712,6 +2713,42 @@ def array_ctypes_to_pointer(context, builder, fromty, toty, val):
     return impl_ret_untracked(context, builder, toty, res)
 
 
+@lower_cast(types.Array, types.Bytes)
+def array_to_bytes(context, builder, fromty, toty, val):
+    arrty = make_array(fromty)
+    arr = arrty(context, builder, val)
+
+    itemsize = arr.itemsize
+    nbytes = builder.mul(itemsize, arr.nitems)
+
+    bstr = _make_constant_bytes(context, builder, nbytes)
+
+    if (fromty.is_c_contig and fromty.layout == "C"):
+        cgutils.raw_memcpy(builder, bstr.data, arr.data, arr.nitems, itemsize)
+    else:
+        shape = cgutils.unpack_tuple(builder, arr.shape)
+        strides = cgutils.unpack_tuple(builder, arr.strides)
+        layout = fromty.layout
+        intp_t = context.get_value_type(types.intp)
+
+        byteidx = cgutils.alloca_once(
+            builder, intp_t, name="byteptr", zfill=True
+        )
+        with cgutils.loop_nest(builder, shape, intp_t) as indices:
+            ptr = cgutils.get_item_pointer2(
+                context, builder, arr.data, shape, strides, layout, indices
+            )
+            srcptr = builder.bitcast(ptr, bstr.data.type)
+
+            idx = builder.load(byteidx)
+            destptr = builder.gep(bstr.data, [idx])
+
+            cgutils.memcpy(builder, destptr, srcptr, itemsize)
+            builder.store(builder.add(idx, itemsize), byteidx)
```

Comment:
This pattern ought to be expressed using a PHI node. But I guess this is not possible due to how `cgutils.loop_nest` is implemented.